### PR TITLE
ADBDEV-5547: Reimplement address availability check for UDP interconnect

### PR DESCRIPTION
**Pull request by bandetto**
Reimplement address availability check for UDP interconnect

This patch reimplements changes introduced in 5beccaa (which backported 70306db,
1ee34a4) and fixes them by correctly handling IPv6 addresses, including
link-local ones.

In these patches, a fix was created for UDPIFC that reworked SendDummyPacket()
by removing the unnecessary call to getaddrinfo(), making
setupUDPListeningSocket() cache the address of the socket when it was first
created.

The original patch (5beccaa) had an oversight in AI_ADDRCONFIG flag usage,
because it does not safeguard against link-local addresses and other non-global
addresses that are automatically assigned upon connecting to certain networks.
These addresses can lead to false positives, even with the AI_ADDRCONFIG flag in
place.

When IPv6 is disabled, all IPv6 addresses are removed from the network
interfaces. However, upon reconnecting to the network, these addresses reappear,
even though IPv6 remains disabled, with a single exception being loopback
interface, IPv6 addresses of which do not reappear.

AI_ADDRCONFIG option is replaced with a manual check to ensure we actually
want to use IPv6, which tries the following:

When IPv6 is initially turned off, the loopback interface also loses its IPv6
address, along with other devices. Unlike actual network interfaces, the
loopback device does not automatically receive a new IPv6 address upon
reconnecting to a network. Therefore, we check for a missing IPv6 address on the
loopback device, because it's a clear sign that IPv6 has been disabled.
Additionally, we check for at least one valid non-link-local IPv6 address as
IPv6 functionality is useless without it. If both of these addresses are
available, we can use the AF_INET6 family.

This is a rework of commit 5beccaa.

---

This PR should be merged by rebasing as three separate commits.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70192

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70222

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70308

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70325

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70340

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70475

**Comment by RekGRpth** [comment]
May be combine last two commits to one, and add following phrase

> This patch should be merged by rebase as three separate commits.

to the end of PR description?

File: src/backend/cdb/motion/ic_udpifc.c [hunk 1]
```diff
@@ -1167,6 +1169,105 @@ resetRxThreadError()
...
+	struct ifaddrs	*ifaddr = NULL;
+	struct ifaddrs	*ifaddrFirst = NULL;
+
+	if (getifaddrs(&ifaddrFirst) == -1)
+	{
+		errorCause = "getifaddrs()";
+		goto end;
+	}
+
+	for (ifaddr = ifaddrFirst; ifaddr != NULL; ifaddr = ifaddr->ifa_next)
```
Comment (RekGRpth) [inline src/backend/cdb/motion/ic_udpifc.c]:
```suggestion
	for (struct ifaddrs *ifaddr = ifaddrFirst; ifaddr != NULL; ifaddr = ifaddr->ifa_next)
```

File: src/backend/cdb/motion/ic_udpifc.c [hunk 2]
```diff
@@ -1167,6 +1169,105 @@ resetRxThreadError()
...
+			const struct sockaddr_in6 *addr =
+				(struct sockaddr_in6 *) ifaddr->ifa_addr;
+			const struct in6_addr *sinAddr = &addr->sin6_addr;
+
+			/* Discard wildcard, and link-local addresses. */
+			if (!IN6_IS_ADDR_UNSPECIFIED(sinAddr) &&
+				!IN6_IS_ADDR_LINKLOCAL(sinAddr))
+			{
+				/* Make sure the address can be used in bind(). */
+				int			bindFd = -1;
```
Comment (RekGRpth) [inline src/backend/cdb/motion/ic_udpifc.c]:
```suggestion
				int			bindFd;
```
or
```suggestion
				int			bindFd = socket(AF_INET6, SOCK_DGRAM, 0);
```

File: src/backend/cdb/motion/ic_udpifc.c [hunk 3]
```diff
@@ -1199,19 +1290,16 @@ setupUDPListeningSocket(int *listenerSocketFd, uint16 *listenerPort, int *txFami
...
-	 * and any connect() or sendto() to this cached address will cause errors.
+	 * Unfortunately, if we are using a wildcard, we can't really distinguish
+	 * whether bind() has really acquired IPv6. Make sure we have at least one
+	 * valid IPv6 address available.
 	 */
-	hints.ai_flags = AI_ADDRCONFIG;
+	hints.ai_family = isIPv6Available() ? AF_UNSPEC : AF_INET;
+
+	hints.ai_socktype = SOCK_DGRAM; /* Datagram socket */
+	hints.ai_protocol = 0;		/* Any protocol */
```
Comment (RekGRpth) [inline src/backend/cdb/motion/ic_udpifc.c]: may be leave these lines unchanged?

File: src/backend/cdb/motion/ic_udpifc.c [hunk 4]
```diff
@@ -1167,6 +1169,105 @@ resetRxThreadError()
...
+
+	if (getifaddrs(&ifaddrFirst) == -1)
+	{
+		errorCause = "getifaddrs()";
+		goto end;
+	}
+
+	for (ifaddr = ifaddrFirst; ifaddr != NULL; ifaddr = ifaddr->ifa_next)
+	{
+		if (ifaddr->ifa_addr != NULL &&
```
Comment (RekGRpth) [inline src/backend/cdb/motion/ic_udpifc.c]: may be use not condition continue here and below to get rid of too nested?

File: src/backend/cdb/motion/ic_udpifc.c (hunk 1 above)
Comment (bandetto) [reply src/backend/cdb/motion/ic_udpifc.c]: That's mixing declarations and code. Every other function in this file declares it's variables at the top of the block, and I followed that style.

File: src/backend/cdb/motion/ic_udpifc.c (hunk 2 above)
Comment (bandetto) [reply src/backend/cdb/motion/ic_udpifc.c]: We don't want to create a socket if the iteration can be skipped by very next if condition. What's wrong with initializing the descriptor to -1?

File: src/backend/cdb/motion/ic_udpifc.c (hunk 4 above)
Comment (bandetto) [reply src/backend/cdb/motion/ic_udpifc.c]: I replaced extra nesting level for link-local/wildcard check with `continue`. If we also remove it here, we'll be unable to use `const` on address pointers.

File: src/backend/cdb/motion/ic_udpifc.c (hunk 2 above)
Comment (RekGRpth) [reply src/backend/cdb/motion/ic_udpifc.c]: > What's wrong with initializing the descriptor to -1?

it is redundant

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70516

File: src/backend/cdb/motion/ic_udpifc.c (hunk 2 above)
Comment (bandetto) [reply src/backend/cdb/motion/ic_udpifc.c]: > it is redundant

I think it does not really matter.

**Review by RekGRpth — DISMISSED** [review]

File: src/backend/cdb/motion/ic_udpifc.c [hunk 5]
```diff
@@ -1167,6 +1169,107 @@ resetRxThreadError()
...
+
+end:
+	errorErrnoSave = errno;
+
+	if (ifaddrFirst != NULL)
+		freeifaddrs(ifaddrFirst);
+
+	if (errorCause != NULL)
+	{
+		errno = errorErrnoSave;
```
Comment (whitehawk) [inline src/backend/cdb/motion/ic_udpifc.c:1256]: Do we really need the `errorErrnoSave`? Do you expect that `freeifaddrs` can modify `errno`?

File: src/backend/cdb/motion/ic_udpifc.c [hunk 6]
```diff
@@ -1167,6 +1169,107 @@ resetRxThreadError()
 }
 
 
+/*
+ * isIPv6Available
+ * 		Check whether a valid IPv6 address exists in the system.
+ *
+ * 		We want to have both loopback IPv6 address, which will show if IPv6 was
```
Comment (whitehawk) [inline src/backend/cdb/motion/ic_udpifc.c]: This statement sounds a bit awkward to me. I suggest to rephrase it to ensure it is not misinterpreted.
For ex.:
```
* To ensure the IPv6 is available in the system, we must check two conditions:
* 1. there is at least one non-link-local IPv6 address available for our use;
* 2. and the loopback IPv6 address exists (as address from item#1 could 
* appear if the interface is reset after disabling of IPv6 in the host system).
```

File: src/backend/cdb/motion/ic_udpifc.c (hunk 5 above)
Comment (bandetto) [reply src/backend/cdb/motion/ic_udpifc.c:1256]: `free()` may change `errno` due to `munmap()` failure before glibc 2.33.

Source:
https://sourceware.org/pipermail/libc-announce/2021/000030.html (2021-02-01):
> The following bugs are resolved with this release:
>
>  ...
>  [17924] malloc: 'free' should not set errno

**Comment by BenderArenadata** [comment]
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1395781

**Review by whitehawk — DISMISSED** [review]

**Comment by BenderArenadata** [comment]
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1395780

**Comment by RekGRpth** [comment]
fix tests please

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70831

**Review by RekGRpth — DISMISSED** [review]

File: src/backend/cdb/motion/ic_udpifc.c [hunk 7]
```diff
@@ -1167,6 +1169,112 @@ resetRxThreadError()
...
+
+			/* Make sure the address can be used in bind(). */
+			bindFd = socket(AF_INET6, SOCK_DGRAM, 0);
+
+			if (bindFd < 0)
+			{
+				if (errno != EAFNOSUPPORT)
+				{
+					errorCause = "socket()";
+					goto end;
```
Comment (andr-sokolov) [inline src/backend/cdb/motion/ic_udpifc.c]: Are "goto end;" and "break" the same?

File: src/backend/cdb/motion/ic_udpifc.c [hunk 8]
```diff
@@ -1167,6 +1169,112 @@ resetRxThreadError()
...
+	for (ifaddr = ifaddrFirst; ifaddr != NULL; ifaddr = ifaddr->ifa_next)
+	{
+		if (ifaddr->ifa_addr != NULL &&
+			ifaddr->ifa_addr->sa_family == AF_INET6)
+		{
+			const struct sockaddr_in6 *addr =
+				(struct sockaddr_in6 *) ifaddr->ifa_addr;
+			const struct in6_addr *sinAddr = &addr->sin6_addr;
+
+			int			bindFd = -1;
```
Comment (andr-sokolov) [inline src/backend/cdb/motion/ic_udpifc.c]: `-1` is not used. This value is replaced with return value of socket()

File: src/backend/cdb/motion/ic_udpifc.c [hunk 9]
```diff
@@ -1167,6 +1169,112 @@ resetRxThreadError()
...
+ *		   if IPv6 is still disabled in the host system.  The loopback device is
+ *		   not affected by such behavior, so a missing IPv6 address on the
+ *		   loopback device is a clear sign that IPv6 has been disabled.
+ */
+static bool
+isIPv6Available(void)
+{
+	bool		hasIpv6Addr = false;
+	bool		hasIpv6Loopback = false;
+	int			errnoSave = errno;
```
Comment (andr-sokolov) [inline src/backend/cdb/motion/ic_udpifc.c]: Do we really need to save errno?

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/71085

**Review by RekGRpth — APPROVED** [review]

**Review by whitehawk — APPROVED** [review]
